contracts/FeeManager.vy

```diff
@@ -0,0 +1,123 @@
+#pragma version 0.3.10
+#pragma optimize gas
+#pragma evm-version shanghai
+
+"""
+@title Compass Fee Manager
+@license MIT
+@author Volume.Finance
+@notice v1.0.0
+"""
+
+interface ERC20:
+    def balanceOf(account: address) -> uint256: view
+    def transfer(to: address, amount: uint256): nonpayable
+
+struct FeeArgs:
+    community_fee: uint256 # Total amount to alot for community wallet
+    security_fee: uint256 # Total amount to alot for security wallet
+    fee_payer_paloma_address: bytes32 # Paloma address covering the fees
+
+compass: public(address)
+grain: public(address)
+
+# Rewards program
+rewards_community_balance: public(uint256) # stores the balance attributed to the community wallet
+rewards_security_balance: public(uint256) # stores the balance attributed to the security wallet
+funds: public(HashMap[bytes32, uint256]) # stores the spendable balance of paloma addresses
+claimable_rewards: public(HashMap[address, uint256]) # stores the claimable balance for eth addresses
+total_funds: public(uint256) # stores the balance of total user funds # Steven: Why do we need this?
+total_claims: public(uint256) # stores the balance of total claimable rewards # Steven: Why do we need this?
+
+@external
+def __init__(_compass: address, grain: address):
+    self.compass = _compass
+    self.grain = grain
+
+@internal
+def compass_check():
+    assert msg.sender == self.compass, "Not Compass"
+
+@external
+@payable
+def deposit(depositor_paloma_address: bytes32):
+    # Deposit some balance on the contract to be used when sending messages from Paloma.
+    # depositor_paloma_address: paloma address to which to attribute the sent amount
+    # amount: amount of COIN to register with compass. Overpaid balance will be sent back.
+    self.compass_check()
+    self.funds[depositor_paloma_address] = unsafe_add(self.funds[depositor_paloma_address], msg.value)
```

Comment:
It can't overflow because the msg.value is actual funds that a user sent and he can't send enormous amount that can overflow uint256.

Reply:
I agree for individual user funds. What about the total values though, like `total_claims` that holds the collective funds from everyone? I imagine it would be pretty hard to create an overflow, but depending on the chain and tokenomics,  it would be possible, right?

Reply:
No, we receive native coin(i.e. ETH on Ethereum and L2s, BNB on BSC chain). And the total minted native coin amount must much less than max(uint256). So even though too many users deposit their funds, the total amount mustn't overflow uint256.

Reply:
Ah, that makes sense. Thanks for explaining!